enterprise/coderd/licenses.go

```diff
@@ -0,0 +1,175 @@
+package coderd
+
+import (
+	"context"
+	"crypto/ed25519"
+	_ "embed"
+	"net/http"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/golang-jwt/jwt/v4"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/coderd"
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/coderd/rbac"
+	"github.com/coder/coder/codersdk"
+)
+
+const (
+	CurrentVersion        = 3
+	HeaderKeyID           = "kid"
+	AccountTypeSalesforce = "salesforce"
+	VersionClaim          = "version"
+	PubSubEventLicenses   = "licenses"
+)
+
+var ValidMethods = []string{"EdDSA"}
+
+//go:embed keys/2022-08-12
+var key20220812 []byte
+
+var keys = map[string]ed25519.PublicKey{"2022-08-12": ed25519.PublicKey(key20220812)}
+
+type Features struct {
+	UserLimit int64 `json:"user_limit"`
+	AuditLog  int64 `json:"audit_log"`
+}
+
+type Claims struct {
+	jwt.RegisteredClaims
+	LicenseExpires *jwt.NumericDate `json:"license_expires,omitempty"`
+	AccountType    string           `json:"account_type,omitempty"`
+	AccountID      string           `json:"account_id,omitempty"`
+	Version        uint64           `json:"version"`
+	Features       Features         `json:"features"`
+}
+
+var (
+	ErrInvalidVersion = xerrors.New("license must be version 3")
+	ErrMissingKeyID   = xerrors.Errorf("JOSE header must contain %s", HeaderKeyID)
+)
+
+// parseLicense parses the license and returns the claims. If the license's signature is invalid or
+// is not parsable, an error is returned.
+func parseLicense(l string, keys map[string]ed25519.PublicKey) (jwt.MapClaims, error) {
+	tok, err := jwt.Parse(
+		l,
+		keyFunc(keys),
+		jwt.WithValidMethods(ValidMethods),
+	)
+	if claims, ok := tok.Claims.(jwt.MapClaims); ok && tok.Valid {
+		version, ok := claims[VersionClaim].(float64)
+		if !ok {
+			return nil, ErrInvalidVersion
+		}
+		if int64(version) != CurrentVersion {
+			return nil, ErrInvalidVersion
+		}
+		return claims, nil
+	}
+	if err != nil {
```

Comment:
Should this err check be immediately after `jwt.Parse`?

Reply:
I think the first `if` and this one are meant to be mutually exclusive, e.g. https://pkg.go.dev/github.com/golang-jwt/jwt#example-ParseWithClaims-CustomClaimsType

Reply:
If err != nil we can’t trust token and this code may do a nil pointer deref. This is the first instance I found: https://github.com/golang-jwt/jwt/blob/4bbdd8ac624fc7a9ef7aec841c43d99b5fe65a29/parser.go#L99

So I’d say err check has to come first.